### PR TITLE
Fix issue #156

### PR DESCRIPTION
**Pull request by metamystical**
Modify DHT.prototype.get to return the value with the greatest sequence number, not the last value retrieved regardless of sequence number

### CODE REVIEW

File: client.js [hunk 1]
```diff
@@ -260,7 +260,7 @@ DHT.prototype.get = function (key, opts, cb) {
       if (!verify || !r.sig || !r.k) return true
       if (!verify(r.sig, encodeSigData(r), r.k)) return true
       if (equals(hash(r.salt ? Buffer.concat([r.salt, r.k]) : r.k), key)) {
-        value = r
+        if(!value || r.seq > value.seq) value = r
```
Comment (feross) [inline client.js]: Can you fix the code style here? There should be a space after `if`. In the future, you can run `npm test` to check this before submitting a PR.

**Comment by feross** [comment]
Deferring to @mafintosh to review this.

**Comment by feross** [comment]
There was a merge conflict after the last PR that I merged, so I fixed it up for you.

**Comment by metamystical** [comment]
Thanks. Sorry about the error.

**Comment by thomasyuan** [comment]
I think this PR is not enough to fix #156 , at least it fails on my test sometimes.
The trick thing for the mutable get is, we shouldn't stop getting value from closest peers even if we already have one because the one we have maybe is not the latest one.

To fix this issue,  for mutable items, we always need return true in onreply(message) to make k-rpc continue to get value from next node. like this:
```c++
    if (isMutable) {
      if (!verify || !r.sig || !r.k) return true
      if (!verify(r.sig, encodeSigData(r), r.k)) return true
      if (equals(sha1(r.salt ? Buffer.concat([r.k, r.salt]) : r.k), key)) {
        console.log('value ', value, 'r.seq', r.seq)
        if (!value || r.seq > value.seq) value = r
        return true  // <=== Here, don't return false.
      }
    } else {

```

**Comment by metamystical** [comment]
I agree with and made the change suggested by thomasyuan above:

> To fix this issue, for mutable items, we always need return true in onreply(message) to make k-rpc continue to get value from next node. like this:
```javascript
    if (isMutable) {
      if (!verify || !r.sig || !r.k) return true
      if (!verify(r.sig, encodeSigData(r), r.k)) return true
      if (equals(sha1(r.salt ? Buffer.concat([r.k, r.salt]) : r.k), key)) {
        console.log('value ', value, 'r.seq', r.seq)
        if (!value || r.seq > value.seq) value = r
        return true  // <=== Here, don't return false.
      }
```

**Comment by thomasyuan** [comment]
Maybe remove the "console.log"? That's for my testing only. Sorry.

**Comment by thomasyuan** [comment]
Oops, the code didn't include that log. Never mind then.

**Comment by metamystical** [comment]
Although this fix works when the DHT nodes and id are persisted, it would be nice if it also worked when the DHT is bootstrapped after a mutable item is stored (put). In that case, the item is usually not found at all. This is because k-rpc stops searching after the first round. If the DHT table contains most of the same nodes used to store the item, one round is enough. But if the table is boostrapped, usually none of the nodes in the table contain the item. I also noticed that k-rpc has trouble finding torrent peers using only the infohash of the torrent and the DHT, for the same reason.

I suspect that the problem is in this part of index.js of k-rpc, in the function _closest (line 211):

```javascript
var closest = table.closest(target, self.k)
if (!closest.length || closest.length < self.bootstrap.length) {
  closest = self.nodes.closest(target, self.k)
  if (!closest.length || closest.length < self.bootstrap.length) bootstrap()
}
```
If `closest.length < self.bootstrap.length` is `true` then `self.nodes` (a K-Bucket) is used to determine the closest nodes to the target unless it is empty (on the first round) in which case it is bootstrapped. The `self.nodes` K-Bucket is built from the client's id, whereas the `table` K-Bucket is built from closest node lists returned by queried nodes. We want to use the `table` K-Bucket to keep the search going. If we use only `self.nodes`, the search will end after one round with no item found, if a node storing the item does not happen to be in the bootstrap list.

If I remove the first `closest.length < self.bootstrap.length`, both problems disappear: I can get a mutable item every time even when I bootstrap the DHT, and I can find torrent peers using only the torrent infohash every time even when I don't bootstrap the DHT.

It looks as though the first `closest.length < self.bootstrap.length` is a mistake because it is used twice. The second instance makes sense because `boostrap()` is called after the conditional. The first instance should probably have a test something like `closest.length < self.nodes.count()` but I'm not sure what the author exactly intended.

**Comment by metamystical** [comment]
While waiting for this pull request to be reviewed, I wrote my own [DHT](https://github.com/metamystical/mdht). It can work with webtorrent.

**Comment by mafintosh** [comment]
Looks good, thanks for the help!

**Comment by mafintosh** [comment]
8.0.0 (with other changes)

**Pull request merged**